### PR TITLE
Add DNS Tests

### PR DESCRIPTION
**Pull request by nlsun**
## High Level Description

This adds tests for the various DNS records generated by Navstar in the various container configurations.

## Related Issues

  - N/A

## Checklist for all PR's

  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [ ] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [ ] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and Review Process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using github's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

### CODE REVIEW

**Comment by orsenthil** [comment]
@nlsun , could you please update the PR template of this PR.

**Comment by nlsun** [comment]
@orsenthil The tests are timing out, please increase the timeouts.

**Comment by orsenthil** [comment]
The error is:

> The build Cloud Integration Tests::DCOS OSS CCM Install Test - API #5721 {buildId=570295} has been running for more than 35 minutes. Terminating...

* If this is consistent, this test is adding a *huge* amount of time. 
* I've bumped it up to 45 minutes, but I am pretty sure that this is not a scalable solution
* And would like to measure how long this **particular test** takes.

**Comment by nlsun** [comment]
@orsenthil @GoelDeepak ready for review.

File: packages/dcos-integration-test/extra/test_service_discovery.py [hunk 1]
```diff
@@ -125,6 +128,206 @@ def _pool_for_mesos_dns():
...
+    return DNSAddresses(container_ips, agent_ips, auto_ips)
+
+
+def get_marathon_addresses_by_service_points(service_points):
+    marathon_host_addrs = frozenset([point.host for point in service_points])
+    marathon_ip_addrs = frozenset([point.ip for point in service_points])
+    return MarathonAddresses(marathon_host_addrs, marathon_ip_addrs)
+
+
+def test_if_service_discovery_works_navstar_mesos_container_host_network(dcos_api_session):
```
Comment (GoelDeepak) [inline packages/dcos-integration-test/extra/test_service_discovery.py]: How about this name? test_if_service_discovery_mesos_host, test_if_service_discovery_mesos_overlay .... and so on

File: packages/dcos-integration-test/extra/test_service_discovery.py [hunk 2]
```diff
@@ -125,6 +128,206 @@ def _pool_for_mesos_dns():
...
+        }
+    }
+    if 'portIndex' in app_definition['healthChecks'][0]:
+        del app_definition['healthChecks'][0]['portIndex']
+    app_definition['healthChecks'][0]['port'] = port
+
+    # Take current definition, strip of the last word "$PORT0" because we
+    # aren't using the port index syntax, and then add on the port we're
+    # actually using.
+    cmd_list = app_definition['cmd'].split()[:-1]
```
Comment (GoelDeepak) [inline packages/dcos-integration-test/extra/test_service_discovery.py]: would it make sense to refactor this in a separate function? Looks like these steps are repeated in other tests as well

File: packages/dcos-integration-test/extra/test_service_discovery.py (hunk 1 above)
Comment (nlsun) [inline packages/dcos-integration-test/extra/test_service_discovery.py]: Yeah, I can make it shorter.

File: packages/dcos-integration-test/extra/test_service_discovery.py (hunk 2 above)
Comment (nlsun) [inline packages/dcos-integration-test/extra/test_service_discovery.py]: Sure

**Comment by nlsun** [comment]
@orsenthil @GoelDeepak I've made the changes

**Comment by nlsun** [comment]
@aeyakovenko you can review for fun as well

**Comment by nlsun** [comment]
Can I actually keep these as separate commits, I think it makes these particular changes easier to follow

**Comment by nlsun** [comment]
adding @drewkerrigan as well

**Comment by GoelDeepak** [comment]
@nlsun I would say reduce the ones that do not have an impact on final version

**Comment by nlsun** [comment]
Fair enough.

File: packages/dcos-integration-test/extra/test_service_discovery.py [hunk 3]
```diff
@@ -125,6 +128,173 @@ def _pool_for_mesos_dns():
...
+# -Container IP
+# -Agent IP
+# -Auto IP
+#
+# More info can be found here: https://dcos.io/docs/1.8/usage/service-discovery/dns-overview/
+
+DNSHost = 0
+DNSPortMap = 1
+DNSOverlay = 2
+DNSAddresses = collections.namedtuple("DNSAddresses", ["container", "agent", "auto"])
```
Comment (orsenthil) [inline packages/dcos-integration-test/extra/test_service_discovery.py]: [minor] newline here at line 153.

File: packages/dcos-integration-test/extra/test_service_discovery.py [hunk 4]
```diff
@@ -125,6 +128,173 @@ def _pool_for_mesos_dns():
...
+    cmd_list = new_app['cmd'].split()[:-1]
+    cmd_list.append(port_str)
+    mesos_cmd = ' '.join(cmd_list)
+    new_app['cmd'] = mesos_cmd
+    return new_app
+
+
+def assert_service_discovery(dcos_api_session, app_definition, net_types):
+    """
+    net_types: List of network types
```
Comment (orsenthil) [inline packages/dcos-integration-test/extra/test_service_discovery.py]: [minor] - Please expand this command. `List of network types: DNSOverlay, DNSPortMap or DNSHost` applicable for navstar.

**Comment by nlsun** [comment]
I've made the changes, just waiting on tests now

**Comment by adam-mesos** [comment]
Why is the new `test_service_discovery_mesos_overlay` failing in Azure?

**Comment by nlsun** [comment]
@adam-mesos 
```
>               'Application deployment failed, reason: {}'.format(data['app']['lastTaskFailure']['message'])
            )
E           AssertionError: Application deployment failed, reason: Executor terminated
```
Looks like the task failed to launch

**Comment by nlsun** [comment]
The difference is that I rebased.

**Comment by orsenthil** [comment]
It looks like that test has not succeed azure ever [TeamCity Link](https://teamcity.mesosphere.io/viewType.html?buildTypeId=ClosedSource_Dcos_IntegrationTests_CloudIntegrationTests_DcosOssAzureIntegrati_2&branch_ClosedSource_Dcos_IntegrationTests_CloudIntegrationTests=pull%2F1271&tab=buildTypeStatusDiv). Other Azure failures were masking it and thus had not really looked at it specifically.

**Comment by adam-mesos** [comment]
Ok, if it never succeeded, then let's pull that test out of the PR so we don't introduce a broken window.

**Comment by nlsun** [comment]
Hmm if that's the case then i prefer to fix the test, looking now.

**Comment by nlsun** [comment]
Ah, looks like there's some underlying flakiness with Azure and the overlay network.

**Comment by nlsun** [comment]
@orsenthil  Jira tracking the issue here: https://jira.mesosphere.com/browse/CORE-893

In the meantime, we can merge+silence that test.

**Comment by orsenthil** [comment]
@nlsun  - I am okay with that. I will like bring the attention of https://jira.mesosphere.com/browse/CORE-893 to @spahl , @adam-mesos and get their buy-in since this new test is **always going to fail** until the underlying mesos+ubuntu bug is fixed.

**Comment by adam-mesos** [comment]
Hmm.. even after disabling nscd, azure still fails on  `test_service_discovery.test_service_discovery_mesos_overlay`

**Comment by nlsun** [comment]
@adam-mesos dug around a bit more... and it turned out nscd was still running. Turns out it's actually called "unscd.service", we disabled "nscd.service". Fixed and pushed, should work now.

**Comment by nlsun** [comment]
Ok, disabling it was not enough as that only goes into effect by the next boot, and the cloud config seems to run _after_ boot, so I've also stopped the service, hopefully it'll work now.

**Comment by nlsun** [comment]
@adam-mesos @orsenthil All green.

**Comment by amitaekbote** [comment]
This failed in three tests on the train, will remove it from this train till fixed

**Pull request merged**